File: 0007.reverse_integer/solution.py
class Solution:
    
    def reverse(self, x):
        # 2**31 == 2147483648 == (2<<30)
        ans = int(('-' if x < 0 else '' ) + str(abs(x))[::-1])
        return 0 if ans < (-2<<30) or ans > (2<<30) else ans

File: 0007.reverse_integer/solution.go
package main

import "fmt"

func reverse(x int) int {
	ans := 0
	for x != 0 {
		pop := x % 10
		// overflow return 0
		if ans > (2<<30)/10 || ans == (2<<30)/10 && pop > 7 {
			return 0
		}

		if ans < (-2<<30)/10 || ans == (-2<<30)/10 && pop < -8 {
			return 0
		}
		ans = ans*10 + pop
		x /= 10
	}
	return ans
}

func main() {
	fmt.Println(reverse(100))
	fmt.Println(reverse(-100))
	fmt.Println(reverse(2<<30 + 1))
	fmt.Println(reverse(-2 << 30))
}
